insights/core/plugins.py

```diff
@@ -96,6 +109,18 @@ def invoke(self, broker):
             log.debug(cpe)
             broker.add_exception(self.component, cpe, traceback.format_exc())
             raise dr.SkipComponent()
+        except TimeoutException as te:
+            log.debug(te)
+            # For some reason when this exception fires, the component is
+            # insights.specs.default.DefaultSpecs.<function> instead of the
+            # actual registry point spec name. So modify the key of the
+            # exception entry to fix it, so it can be accessed properly
+            # during dehydration.
+            broker.add_exception("insights.specs.Specs.{0}".format(
+                dr.get_name(self.component).split('.')[-1]), te, traceback.format_exc())
```

Comment:
Here, for datasources, the `self.component` would be `insights.specs.datasources.<module>.<function>`, based on the current code, the key inserted to the `broker.exceptions` is `insights.specs.Specs.<function>`.

But this `insights.specs.Specs.<function>` will never be a valid name when doing the dehydration.  https://github.com/RedHatInsights/insights-core/blob/e713e3f5814da3580622ad919d836bc5dd7b7056/insights/core/serde.py#L202

The `dehydrate()` will only dehydrate the specs defined under the `insights.specs.Specs` module according to the "persist" items declared in the `default_mainfest` (currently, `insights.specs.Specs.<specs>` only)

We may need to get the original `spec` (defined in the `insights.specs.Specs.<spec>`) of this `self.component` and depends on the `self.component` via itself. e.g.:
```python
            def _get_spec(comp):
                comps = dr.get_dependents(comp)
                for cmp in comps:
                    sp_name = dr.get_name(cmp).split('.')
                    if len(sp_name) == 4 and sp_name[2] == "Specs":
                        return cmp
                    return _get_spec(cmp)

            broker.add_exception(dr.get_name(_get_spec(self.component)), te, traceback.format_exc())
```

Reply:
Ideally, to make it strict, I think we may need to ensure it is consistent with the "persist", e.g. only insert the `traceback` when it's declared in the `persist`  (`insights.specs.Specs` by default).

Reply:
So the issue I was running into was the exceptions weren't being dehydrated in serde when datasources are called directly. This is because the key it was using to check was insights.specs.Specs.system_user_dirs for example, but the exception entry key was the function itself. So at first I changed the entry key when adding the exception to the name of the component, but that came out to be insights.specs.default.DefaultSpecs.system_user_dirs. I didn't know of a way to query the actual spec name. But thank you for the get_dependents function, I altered the _get_spec suggested function a little bit and using it works much better than what I had.

Reply:
The current `_get_spec_name` works well when the name of datasource is the same as the `symbolic_name` of the spec, e.g. the `system_user_dirs`. 
https://github.com/RedHatInsights/insights-core/blob/9a6b5c37ed9f69d798946922c8cfbbf50e7e0a0d/insights/specs/__init__.py#L702
https://github.com/RedHatInsights/insights-core/blob/9a6b5c37ed9f69d798946922c8cfbbf50e7e0a0d/insights/specs/default.py#L617

But it won't work when they are different, e.g. the `du_dirs` :

https://github.com/RedHatInsights/insights-core/blob/9a6b5c37ed9f69d798946922c8cfbbf50e7e0a0d/insights/specs/__init__.py#L150
https://github.com/RedHatInsights/insights-core/blob/9a6b5c37ed9f69d798946922c8cfbbf50e7e0a0d/insights/specs/default.py#L175

In this case of `du_dirs`, the `_get_spec_name` returns `insights.specs.default.DefaultSpecs.du_dirs`, but `insights.specs.Specs.du_dirs` is expected.

And it seems the `du_dirs` is not the only case.

Reply:
Good catch, I fixed it by returning the name of the component if ".Specs" isn't found. For some reason datasources called from a spec factory function like foreach_execute never had the weird key issue.